[Sema] Warn on unsound pointer conversions

hamishknight opened the pull request:
In this PR is the logic to emit a warning on unsound array-to-pointer, string-to-pointer, and inout-to-pointer conversions. This is done through the introduction of an underscored attribute `@_nonEphemeral` which denotes a pointer parameter which cannot accept pointer conversions that are only valid for the duration of the call they're passed to.

This includes all array-to-pointer and string-to-pointer conversions, and most inout-to-pointer conversions with the exception of:
- Global and static variables
- Stored properties on such variables that are accessed directly (e.g excluding ones with property observers, ones on non-final classes and resilient types in general)
- Force unwraps of any of the above

Various pointer parameters in the stdlib have been marked as `@_nonEphemeral`, primarily pointer parameters for initialisers on pointer types. In addition, this PR infers `@_nonEphemeral` on pointer parameters for memberwise initialisers and enum case constructors.

Essentially, this is the warning logic of #20070 plus:
- ~~The allowing of stored properties on final class global/static vars to be treated as non-ephemeral.~~
- An additional note that suggests that the user use `withUnsafe[Mutable][Bytes/Pointer]`, or for arrays, `withUnsafe[Mutable][Bytes/BufferPointer]`, or for strings, `withCString`.
- The inferring of `@_nonEphemeral` for enum case constructors.
- More tests.

If possible, I'd like to get this into Swift 5, allowing us to transition to an error for Swift 6. I do appreciate how late we are into the release cycle for such a (relatively big) change though.

I'm unsure if this PR requires Swift evolution discussion – I'm happy to start up a thread if it's felt that it does.

Resolves [SR-2790].

  [SR-2790]: https://bugs.swift.org/browse/SR-2790

hamishknight committed:
[stdlib] Mark some parameters @_nonEphemeral

These include the pointer-to-pointer and pointer-to-buffer-pointer initialiser parameters amongst a couple of others, such as `Unmanaged.fromOpaque`, and the source for the `move[...]` family of methods.

hamishknight committed:
[Sema] Warn on unsound pointer conversions

Warn on ephemeral conversions that are passed to @_nonEphemeral parameters. This will hopefully transition to an error in a future version.

hamishknight committed:
[AST] Add Unsafe[Mutable][Raw]Buffer types as known decls

hamishknight committed:
[AST] Add @_nonEphemeral type attribute

This non-user-facing attribute is used to denote pointer parameters which do not accept pointers produced from temporary pointer conversions such as array-to-pointer, string-to-pointer, and in some cases inout-to-pointer.

hamishknight committed:
[Synthesis] Infer @_nonEphemeral for various pointer parameters

These include memberwise initializers for pointer properties and enum case constructors for pointer payloads. In both cases, the pointer is being immediately escaped, so don't allow the user to pass a temporary pointer value.

hamishknight committed:
Allow stored properties on final class global and static vars to be non-ephemeral

For example, this enables:

```swift
final class C {
  var foo: Int = 0
}
var c = C()
let ptr = UnsafePointer(&c.foo)
```

hamishknight committed:
Add a note to the non-ephemeral diagnostic that suggests an alternative

- For a string, we suggest withCString
- For an array, we suggest withUnsafe[Mutable][Bytes/BufferPointer]
- For an arbitrary inout-to-pointer, we suggest withUnsafe[Mutable][Bytes/Pointer]

hamishknight commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -365,6 +365,35 @@ ERROR(cannot_convert_argument_value_generic,none,
+WARNING(cannot_convert_non_ephemeral_warning,none,
+        "passing temporary pointer argument of type %0 to parameter expecting "
+        "pointer that outlives the duration of the call leads to undefined "
+        "behaviour; this will be an error in a future release", (Type))
```
This diagnostic feels a bit too wordy, but I can't think of how to improve it – any ideas @jrose-apple?

jrose-apple replied:
Hm, well, we could remove that "u" from "behavior". :-)

> warning: temporary pointer will be destroyed after call to %0

Okay, that sounds sufficiently scary, but the problem is that the user probably doesn't think they've written a pointer. They wrote an inout expression or an array or string.

> error: cannot pass array to %0; argument %1 must be a pointer that outlives the call to %0
> error: cannot pass string to %0; argument %1 must be a pointer that outlives the call to %0
> error: cannot use inout expression here; argument %1 must be a pointer that outlives the call to %0

How does that sound? And then we get the warning form for pre-Swift-5 (which we maybe take away completely in the future):

> warning: passing array to %0, but argument %1 should be a pointer that outlives the call to %0
> warning: passing string to %0, but argument %1 should be a pointer that outlives the call to %0
> warning: inout expression creates a temporary pointer, but argument %1 should be a pointer that outlives the call to %0

I don't love this but it's a little less wordy?

hamishknight replied:
Oops, need to start using a US spell checker :) Thanks for the suggestions, I like them!

IMO it would be nice to use the argument type of the array or string in the diagnostic message, and avoid repeating the name of the callee – so how about the following?

> error: cannot pass %0 to parameter; argument must a pointer that outlives the call%select{| to %2}1

> warning: passing %0 to parameter, but argument should be a pointer that outlives the call%select{| to %2}1; this will be an error in a future release

> error: cannot use inout expression here; argument must a pointer that outlives the call%select{| to %1}0

> warning: inout expression creates a temporary pointer, but argument should be a pointer that outlives the call%select{| to %1}0; this will be an error in a future release

For now, I've omitted the placeholder for "argument %1", as I wasn't too sure what it would be filled in with – would it be the argument position or the text of the argument expr?

I've gone ahead and pushed a commit with the message changes so far.

jrose-apple replied:
Typo: "must a pointer".

The reason I had the name of the callee twice was that "to parameter" on its own sounds weird to me. The "argument %1" part was the position in my head, to make it slightly easier to tell what the diagnostic is referring to when just looking at the line and not the column or highlight.

I feel like the "this will be an error in a future release" isn't really adding that much value either, especially for a message that's already fairly long.

hamishknight replied:
> Typo: "must a pointer"

Oops, good catch!

> The reason I had the name of the callee twice was that "to parameter" on its own sounds weird to me.

Interesting! "to parameter" sounds natural to me – nonetheless I'm happy to change it. Though for cases where we don't have a callee decl, should it be "to parameter" or "to call"? The former sounds more natural to me, but the latter is more consistent with the message when we do have a callee.

> The "argument %1" part was the position in my head, to make it slightly easier to tell what the diagnostic is referring to when just looking at the line and not the column or highlight.

Thanks for clarifying, I'll go ahead and make that change.

> I feel like the "this will be an error in a future release" isn't really adding that much value either, especially for a message that's already fairly long.

Fair enough – I added it because I felt the diagnostic wasn't quite scary enough without it, but given the length of the message, I agree it probably doesn't quite pull its weight.

hamishknight committed:
Amend non-ephemeral diagnostic message in response to feedback

atrick reviewed (commented):
I'll let @jrose-apple approve this because I don't know Sema well enough to determine if you're missing anything. But I did review the code and tests, and they looks astonishingly thorough to me.

hamishknight commented on test/Sema/diag_non_ephemeral_swift4.swift:
```diff
@@ -0,0 +1,448 @@
+  }
+}
+
+// We can also project stable and, in the case of final classes, semi-stable pointer values
```
Ah sorry, I should have expanded on that a bit more – what I meant was "stable until the object is deallocated". I'll edit the comment.

Thanks for looking over the PR!

hamishknight committed:
Address review feedback

- Add/clarify a couple of comments
- Add a couple of tests
- Rename test file
- Remove extraneous locator simplification logic

jrose-apple commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2356,3 +2356,110 @@ Expr *constraints::simplifyLocatorToAnchor(ConstraintSystem &cs,
+  case ConversionRestrictionKind::SetUpcast:
+  case ConversionRestrictionKind::HashableToAnyHashable:
+  case ConversionRestrictionKind::CFTollFreeBridgeToObjC:
+  case ConversionRestrictionKind::ObjCTollFreeBridgeToCF:
```
Should this be llvm_unreachable, then?

hamishknight replied:
In this patch it could be, although IMO it should it be reasonable for the function to take an arbitrary conversion and answer the question "is this something that we could pass to a `@_nonEphemeral` parameter?" – which is true for the rest of these conversions, as they're not things that `@_nonEphemeral` restricts.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -8403,3 +8412,83 @@ bool swift::diagnoseBaseUnwrapForMemberAccess(Expr *baseExpr, Type baseType,
+                      : diag::cannot_construct_dangling_pointer;
+
+    diags.diagnose(anchor->getLoc(), diagID, instanceTy, constructorKind)
+        .highlight(anchor->getSourceRange());
```
Is the anchor expression always the correct expression? What if the construction is nested inside a larger expression?

hamishknight replied:
As I understand it, unsimplified argument locators for an apply/subscript are always anchored on the apply/subscript expr that they correspond to.

This is how they're generated in CSApply and CSGen at least:

https://github.com/apple/swift/blob/af1391400940b41ee50d4766405aa538133b279a/lib/Sema/CSGen.cpp#L2561-L2564

https://github.com/apple/swift/blob/af1391400940b41ee50d4766405aa538133b279a/lib/Sema/CSApply.cpp#L3085-L3089

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -161,6 +161,15 @@ void constraints::simplifyLocator(Expr *&anchor,
+        anchor = subscriptExpr->getIndex();
+        path = path.slice(1);
+        continue;
+      }
```
What else is this going to affect?

hamishknight replied:
From what I could tell, the majority of usages of `simplifyLocator` occur within CSDiag, and the usages outside of CSDiag appear to revolve around locators to decls, which this shouldn't affect.

So as far as I can tell, the worst-case impact would be a slight change in diagnostics (but even then, it should be a change for the better, as we're now correctly simplifying a subscript argument locator to the argument expr).

That being said, looking at it again, it appears this logic isn't necessary for this patch, so I'll remove it to minimise risk.

jrose-apple commented on lib/Sema/TypeCheckPattern.cpp:
```diff
@@ -838,6 +838,16 @@ bool TypeChecker::typeCheckParameterList(ParameterList *PL,
+          unwrappedType = objectType;
+        if (unwrappedType->getAnyPointerElementType())
+          param->setNonEphemeral();
+      }
```
Do we also want this for stored properties?

hamishknight replied:
Yes, and we're inferring `@_nonEphemeral` for parameters of memberwise initialisers – did you have another case in mind?

jrose-apple replied:
I meant setters, sorry.

hamishknight replied:
Ah, I see! Though I can't immediately think of a case where you can use X-to-pointer conversions in conjunction with setters, AFAIK we only allow them for arguments of applies/subscripts.

jrose-apple replied:
Ah, I didn't realize we didn't perform those conversions for assignments! Okay, good.

```swift
struct Test {
  var ptr: UnsafeRawPointer?
}

var x = Test()
x.ptr = "abc" // already an error
```

jrose-apple commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2417,14 +2404,22 @@ bool ConstraintSystem::isConversionNonEphemeral(
 
+      // Look through load exprs. These can occur when projecting through a
+      // property on a class, which we want to support if the class is
+      // non-final.
```
I don't think this is the correct condition. What case is this trying to handle?

hamishknight replied:
This is handling the following (which we want to allow):

```swift
final class C {
  var str = ""
}

struct S {
  var c = C()
}

var s = S()
let ptr = UnsafeMutablePointer(&s.c.str)
```

The base `s` is an lvalue, but we want an rvalue `C`, so a `LoadExpr` is inserted before then performing the member access for `str`.

jrose-apple replied:
I don't think we want to allow this. Basically if you load from a class, you don't care what the base was anymore, and then either we allow forming pointers to instance variables or we don't. Being a global doesn't matter at that point. Right, @atrick?

atrick replied:
`let ptr = UnsafeMutablePointer(&s.c.str)`

If `s` is a `var`, then there's there's nothing keeping `c` alive. So escaping the addres of `c.str` seems pretty bad.

If `s` is a top-level or global `let` then `c` stays alive. However, technically we don't guarantee that a Swift class object has a persistent address (it may need to for satisfying some runtime constraint, but the language doesn't guarantee it). So I don't think we should allow that either.

Sorry if I missed that in my review.

atrick replied:
Looking through `diag_non_ephemeral_warning.swift` again, I see these two cases that still don't produce warnings, but probably should:
```
takesRaw(&topLevelFinalC.storedProperty)
takesRaw(&topLevelFragileFinalC.storedProperty)
```

jrose-apple replied:
I think we probably do want to make that guarantee, but even without that differentiating between global `let` and global `var` seems way too subtle.

atrick replied:
We do support `realloc` on uniquely reference objects.

hamishknight replied:
@jrose-apple As I understand it, the base still matters for class instances. For example, we don't want to allow users to form a pointer to an instance variable on a class-typed local variable, as we don't guarantee it will remain initialised until the end of the stack frame.

hamishknight replied:
> `let ptr = UnsafeMutablePointer(&s.c.str)`
>
> If `s` is a `var`, then there's there's nothing keeping `c` alive. So escaping the addres of `c.str` seems pretty bad.

@atrick Should `c` not be guaranteed to be kept alive as long as `s.c` isn't changed to a different reference? IMO this would be reasonable constraint to place on the lifetime of the pointer created from the inout-to-pointer conversion.

> If `s` is a top-level or global `let` then `c` stays alive. However, technically we don't guarantee that a Swift class object has a persistent address (it may need to for satisfying some runtime constraint, but the language doesn't guarantee it). So I don't think we should allow that either.

Ah, okay then. I think for now at least I'll revert b6e1e996bd3c99e70c7c344d4c2ea5010cea4eb9, with the view of possibly reinstating it if we ever officially make such a guarantee. I somewhat doubt many (if any) people are using this pattern anyway – there was no sign of it in the source compat suite.

atrick replied:
> @atrick Should c not be guaranteed to be kept alive as long as s.c isn't changed to a different reference? IMO this would be reasonable constraint to place on the lifetime of the pointer created from the inout-to-pointer conversion.

Yes. If `s` is a `let`, then we have that guarantee. And on second thought, I don't think it will ever be possible to reallocate an object that has a `let` reference. But I think @jrose-apple and I agree that it's better to have a simple rule that disallows non-ephemeral pointers to instance variables.

atrick replied:
> Ah, okay then. I think for now at least I'll revert b6e1e99,

Wait a sec, that commit also allows non-ephemeral pointers to `static` variables. I think that's ok under the circumstances that you've identified.

atrick replied:
> Does this include non-ephemeral pointers to stored instance properties on global/static variable structures?

No, stored properties on structs can be pointed to if the struct itself is global/static. We want to disallow pointing to stored properties inside class instances, which I think we were calling "instance variables".

jrose-apple commented on lib/Sema/CSApply.cpp:
```diff
@@ -5475,22 +5482,31 @@ static void checkNonEphemeralArgumentConversion(
     diagnoseIllegalEphemeralConversion(
-        cs.getASTContext(), *conversion, argExpr, argType, param.getPlainType(),
-        callee, fnType, argLocator.getAnchor(),
+        cs.getASTContext(), *conversion, argIdx, subExpr, argType,
```
Now that you're passing `subExpr`, I don't think you need to pass `argType`.

hamishknight replied:
`diagnoseIllegalEphemeralConversion` will eventually also want to be able to take expressions that haven't had solutions applied to them, so I *believe* the type needs to be passed separately (at least, I'm not sure whether it's safe to rely on CSApply giving us back the same `InOutExpr` such that we can query the type from the constraint system).

jrose-apple commented on test/Sema/diag_non_ephemeral_swift4.swift:
```diff
@@ -21,46 +21,46 @@ var optionalArr: [Int8]?
 
-takesMutableRaw(&arr, 5) // expected-warning {{passing temporary pointer argument of type 'UnsafeMutableRawPointer' to parameter expecting pointer that outlives the duration of the call leads to undefined behaviour; this will be an error in a future release}}
-// expected-note@-1 {{implicit argument conversion from '[Int8]' to 'UnsafeMutableRawPointer' produces a pointer valid only for the duration of the call}}
+takesMutableRaw(&arr, 5) // expected-warning {{inout expression creates a temporary pointer, but argument #1 should be a pointer that outlives the call to 'takesMutableRaw'}}
```
It's weird to have a diag_non_ephemeral_swift4.swift without the corresponding diag_non_ephemeral_swift5.swift.

hamishknight replied:
Yeah, that's a holdover from when I had the logic combined for a warning in Swift 4 mode and error for Swift 5 mode. I'll rename it to `diag_non_ephemeral_warning`.

Thanks for the review!

jrose-apple reviewed (commented):
There's a lot here so I might have still missed something, but this is what I noticed reading through the whole thing. @xedin or @rudkx should probably comment on the ConstraintSystem parts.

hamishknight commented on test/attr/attributes.swift:
```diff
@@ -273,3 +273,29 @@ class SILStored {
+
+protocol P {}
+extension P {
+  // Allow @_nonEphemeral on the protocol Self type, as the protocol could be adopted by a pointer type.
```
You're right, it's not technically required, but it's just something I thought would be nice. It's not entirely foolproof either, as you could mark a protocol `Self` parameter as `@_nonEphemeral` and then have the protocol adopted by a non-pointer type. I'm happy to rip out the check if you want.

hamishknight replied:
We need this for the two initialisers defined on the stdlib's `_Pointer` protocol, which take `Self(?)` parameters and can accept X-to-pointer conversions:

https://github.com/apple/swift/blob/af1391400940b41ee50d4766405aa538133b279a/stdlib/public/core/Pointer.swift#L79-L95

jrose-apple replied:
Ugh. Okay.

xedin commented:
I'll try to take a look at Sema changes tomorrow.

hamishknight committed:
Revert "Allow stored properties on final class global and static vars to be non-ephemeral"

b6e1e996bd3c99e70c7c344d4c2ea5010cea4eb9

hamishknight committed:
Merge remote-tracking branch 'upstream/master' into health-and-safety-pointers-ep1

hamishknight committed:
Revert ASTMangler hack

This was fixed in 99c78894efa8342f12ead097c703cd01b11014e4.

hamishknight commented on include/swift/AST/Decl.h:
```diff
@@ -4854,11 +4862,15 @@ class ParamDecl : public VarDecl {
 
-  /// The default value, if any, along with whether this is varargs.
-  llvm::PointerIntPair<StoredDefaultArgument *, 1> DefaultValueAndIsVariadic;
+  enum class Flags {
```
Sure thing! I've opened a separate PR: https://github.com/apple/swift/pull/20663

xedin commented on lib/Sema/CSApply.cpp:
```diff
@@ -5454,6 +5459,59 @@ static bool hasCurriedSelf(ConstraintSystem &cs, ConcreteDeclRef callee,
   return false;
 }
 
+static void checkNonEphemeralArgumentConversion(
```
I think it would make most sense to make it a`TypeMatchFlag`. Maybe some constraints should just carry type match options which could be deduced by `matchCallArguments`? Or otherwise it has to be a "constraint restriction"...

hamishknight commented:
Sorry for taking so long to get back to this! I have opened an updated PR here: https://github.com/apple/swift/pull/27695.
